fetch/connection-pool/network-partition-key.html

```diff
@@ -0,0 +1,266 @@
+    popup_params: [
+      {type: 'main_frame'},
+      {type: 'main_frame'}
+    ]
+  },
+  {
+    name: 'Cross-site resources with credentials',
+    subresource_origin: OTHER_ORIGIN,
+    include_credentials: true,
+    popup_params: [
+      {type: 'main_frame'},
+      {type: 'main_frame'}
+    ]
+  },
+  {
+    name: 'Cross-site resources without credentials',
+    subresource_origin: OTHER_ORIGIN,
+    include_credentials: false,
+    popup_params: [
+      {type: 'main_frame'},
+      {type: 'main_frame'}
+    ]
+  },
+  {
+    name: 'Iframes',
+    subresource_origin: OTHER_ORIGIN,
+    include_credentials: true,
+    popup_params: [
+      {
+        type: 'iframe',
+        iframe_origin: OTHER_ORIGIN
+      },
+      {
+        type: 'iframe',
+        iframe_origin: OTHER_ORIGIN
+      }
+    ]
+  },
+  {
+    name: 'Workers',
+    subresource_origin: POPUP_ORIGINS[0],
+    include_credentials: true,
+    popup_params: [
+      {type: 'worker'},
+      {type: 'worker'}
+    ]
+  },
+  {
+    name: 'Workers with cross-site resources',
+    subresource_origin: OTHER_ORIGIN,
+    include_credentials: true,
+    popup_params: [
+      {type: 'worker'},
+      {type: 'worker'}
+    ]
+  },
+  {
+    name: 'CSP sandbox',
+    subresource_origin: POPUP_ORIGINS[0],
+    include_credentials: true,
+    popup_params: [
+      {type: 'csp_sandbox'},
+      {type: 'csp_sandbox'}
+    ]
+  },
+  {
+    name: 'about:blank from opaque origin iframe',
+    subresource_origin: OTHER_ORIGIN,
+    include_credentials: true,
+    popup_params: [
+      {type: 'opaque_about_blank'},
+      {type: 'opaque_about_blank'}
+    ]
+  },
+];
+
+const BASE_PATH = window.location.pathname.replace(/\/[^\/]*$/, '/');
+
+function create_script_url(origin, uuid, partition_id, dispatch) {
+  return `${origin}${BASE_PATH}resources/network-partition-key.py?uuid=${uuid}&partition_id=${partition_id}&dispatch=${dispatch}`
+}
+
+function run_test(test) {
+  var uuid = token();
+
+  // Used to track the opened popup windows, so they can be closed at the end of the test.
+  // They could be closed immediately after use, but safest to keep them open, as browsers
+  // could use closing a window as a hint to close idle sockets that the window used.
+  var popup_windows = [];
+
+  // Creates a popup window at |url| and waits for a test result. Returns a promise.
+  function create_popup_and_wait_for_result(url) {
+    return new Promise(function(resolve, reject) {
+      popup_windows.push(window.open(url));
+      // Listen for the result
+      function message_listener(event) {
+        if (event.data.result === "success") {
+          resolve();
+        } else if (event.data.result === "error") {
+          reject(event.data.details);
+        } else {
+          return; // Ignore testharness.js internal messages
+        }
+        window.removeEventListener("message", message_listener);
+      }
+      window.addEventListener("message", message_listener);
```

Comment:
You can also use `{ once: true }` as third parameter if you expect only a single call.

Reply:
> I found a couple of nits and it would probably be good if someone else took a look as well. One thing I didn't quite understand is how connections are distinguished server side.
> 
> The other thing that's not immediately clear to me is if this can even be tested in H/1. I had the impression that each request/response pair got its own connection there.

Thanks, Anne!

That's mostly true.  Connections aren't reused for responses that don't include a Content-Length response header.  Responses produced by running *.py files, fortunately, all include a Content-Length header.  Responses that are just the contents of some file do not.  So using as few non-Python responses as possible makes the test more robust, since it prevents sockets from being closed by the server.  We could look into changing behavior for non-Python responses, but decided it wasn't worth the effort, and I didn't want to change the current behavior of a huge number of tests..

I did test that these test pass in Chrome with connection partitioning, and fail without it (Except one of the opaque origin tests, as there's a Chrome bug that needs a fair bit of refactoring before it can be fixed)

Reply:
Done

Reply:
> You can also use `{ once: true }` as third parameter if you expect only a single call.

Thanks, Anne!

That's mostly true. Connections aren't reused for responses that don't include a Content-Length response header. Responses produced by running *.py files, fortunately, all include a Content-Length header. Responses that are just the contents of some file do not. So using as few non-Python responses as possible makes the test more robust, since it prevents sockets from being closed by the server. We could look into changing behavior for non-Python responses, but decided it wasn't worth the effort, and I didn't want to change the current behavior of a huge number of tests..

I did test that these test pass in Chrome with connection partitioning, and fail without it (Except one of the opaque origin tests, as there's a Chrome bug that needs a fair bit of refactoring before it can be fixed)